File: data_structures/arrays_and_strings/are_perms.js
function arePerms(str1, str2) {
  if (str1.length !== str2.length) {
    return false;
  }
  let chars1 = str1.split("");
  let chars2 = str2.split("");
  let counts1 = {};
  let counts2 = {};
  
  for (let i = 0; i < chars1.length; i++) {
    if (counts1[chars1[i]]) {
      counts1[chars1[i]] += 1;
    } else {
      counts1[chars1[i]] = 1;
    }
  }
  
  for (let i = 0; i < chars2.length; i++) {
    if (counts2[chars2[i]]) {
      counts2[chars2[i]] += 1;
    } else {
      counts2[chars2[i]] = 1;
    }
  }
  
  for (const key in counts1) {
    if (counts1[key] !== counts2[key]) {
      return false;
    }
  }
  
  return true;
}

// Time complexity: O(n)
// Space complexity: O(n)
// This is pretty lengthy but avoids sorting O(n log n)

File: data_structures/arrays_and_strings/are_perms.rb
def are_perms?(str1, str2)
  chars1 = Hash.new(0)
  chars2 = Hash.new(0)
  str1.each_char do |ch|
    chars1[ch.downcase] += 1
  end
  str2.each_char do |ch|
    chars2[ch.downcase] += 1
  end
  chars1 == chars2
end

# Time complexity: O(n)
# Space complexity: O(n)
# Initial mistakes: Forgot to account for uppercase, need to ask interviewer first
# So it looks like in Ruby, you can do direct comparisons of arrays
# and hashes and it will be fine.